ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -93,2565 +92,1438 @@
+ * to converting halfwidth to fullwidth.
+ * However, this function can also convert in the opposite direction, or convert
+ * hiragana to katakana and so on... all depending on the value of the `mode` argument */
+mbfl_string *mbfl_ja_jp_hantozen(mbfl_string *string, mbfl_string *result, int mode)
 {
-	size_t n;
-	unsigned char *p;
 	mbfl_memory_device device;
-	mbfl_convert_filter *decoder = NULL;
-	mbfl_convert_filter *encoder = NULL;
-	mbfl_convert_filter *tl_filter = NULL;
-	mbfl_convert_filter *next_filter = NULL;
-	mbfl_filt_tl_jisx0201_jisx0208_param *param = NULL;
 
 	mbfl_memory_device_init(&device, string->len, 0);
-	mbfl_string_init(result);
+	mbfl_string_init_set(result, string->encoding);
 
-	result->encoding = string->encoding;
+	mbfl_convert_filter *decoder = mbfl_convert_filter_new(&mbfl_encoding_wchar,
+		string->encoding, mbfl_memory_device_output, 0, &device);
 
-	decoder = mbfl_convert_filter_new(
-		&mbfl_encoding_wchar,
-		string->encoding,
-		mbfl_memory_device_output, 0, &device);
-	if (decoder == NULL) {
-		goto out;
-	}
-	next_filter = decoder;
-
-	param = emalloc(sizeof(mbfl_filt_tl_jisx0201_jisx0208_param));
+	/* TODO: should this not be part of the filter ctor?? */
+	mbfl_filt_tl_jisx0201_jisx0208_param *param = emalloc(sizeof(mbfl_filt_tl_jisx0201_jisx0208_param));
 	param->mode = mode;
 
-	tl_filter = mbfl_convert_filter_new2(
-		&vtbl_tl_jisx0201_jisx0208,
-		(int(*)(int, void*))next_filter->filter_function,
-		(int(*)(void*))next_filter->filter_flush,
-		next_filter);
-	if (tl_filter == NULL) {
-		efree(param);
-		goto out;
-	}
-
+	mbfl_convert_filter *tl_filter = mbfl_convert_filter_new2(&vtbl_tl_jisx0201_jisx0208,
+		(filter_output_func)decoder->filter_function, decoder->filter_flush, decoder);
 	tl_filter->opaque = param;
-	next_filter = tl_filter;
-
-	encoder = mbfl_convert_filter_new(
-		string->encoding,
-		&mbfl_encoding_wchar,
-		(int(*)(int, void*))next_filter->filter_function,
-		(int(*)(void*))next_filter->filter_flush,
-		next_filter);
-	if (encoder == NULL) {
-		goto out;
-	}
 
-	/* feed data */
-	p = string->val;
-	n = string->len;
-	if (p != NULL) {
-		while (n > 0) {
-			if ((*encoder->filter_function)(*p++, encoder) < 0) {
-				break;
-			}
-			n--;
-		}
-	}
+	mbfl_convert_filter *encoder = mbfl_convert_filter_new(string->encoding, &mbfl_encoding_wchar,
+		(filter_output_func)tl_filter->filter_function, tl_filter->filter_flush, tl_filter);
 
+	mbfl_filter_feed_data(string, encoder);
 	mbfl_convert_filter_flush(encoder);
-	result = mbfl_memory_device_result(&device, result);
-out:
-	if (tl_filter != NULL) {
-		if (tl_filter->opaque != NULL) {
-			efree(tl_filter->opaque);
-		}
-		mbfl_convert_filter_delete(tl_filter);
-	}
+	mbfl_memory_device_result(&device, result);
 
-	if (decoder != NULL) {
-		mbfl_convert_filter_delete(decoder);
-	}
-
-	if (encoder != NULL) {
-		mbfl_convert_filter_delete(encoder);
+	if (tl_filter->opaque != NULL) {
```

Comment:
Almost all the `mbfl_convert_filter` structs could be stack-allocated for sure. I was actually thinking about doing that, but haven't done it yet, since it doesn't show as a significant cost when profiling.

The NULL check disappears in a later commit from this PR.